packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -0,0 +1,591 @@
+    ) {
+        let inner = self.inner.get_or_insert_with(BTreeMap::new);
+
+        inner.insert(
+            TypeId::of::<T>(),
+            Box::new(move |frame, ctx| {
+                // `.unwrap()` never fails here, because `Hooks` guarantees the function will never
+                // be called on an object which cannot be downcast.
+                let frame = frame.downcast_ref::<T>().unwrap();
+
+                hook(frame, ctx.cast())
+            }),
+        );
+    }
+
+    pub(crate) fn fallback(
+        &mut self,
+        hook: impl for<'a> Fn(&Frame, &mut HookContext<'a, Frame>) -> Option<Emit>
+        + Send
+        + Sync
+        + 'static,
+    ) {
+        self.fallback = Some(Box::new(hook));
+    }
+
+    pub(crate) fn call<'a>(&self, frame: &Frame, ctx: &mut HookContext<'a, Frame>) -> Option<Emit> {
+        let ty = Frame::type_id(frame);
+
+        if let Some(hook) = self.inner.as_ref().and_then(|map| map.get(&ty)) {
+            Some((hook)(frame, ctx))
+        } else if let Some(fallback) = self.fallback.as_ref() {
+            (fallback)(frame, ctx)
+        } else {
+            builtin(frame, ctx)
+        }
+    }
+}
+
+mod default {
+    #![allow(unused_imports)]
+
+    #[cfg(any(all(nightly, feature = "std"), feature = "spantrace"))]
+    use alloc::format;
+    #[cfg(all(nightly, feature = "std"))]
+    use std::backtrace::Backtrace;
+
+    #[cfg(feature = "spantrace")]
+    use tracing_error::SpanTrace;
+
+    use crate::{
+        fmt::{hook::HookContext, Emit},
+        Frame,
+    };
+
+    #[cfg(all(nightly, feature = "std"))]
+    fn backtrace(backtrace: &Backtrace, ctx: &mut HookContext<Backtrace>) -> Emit {
+        let idx = ctx.increment();
+
+        ctx.attach_snippet(format!("Backtrace No. {}\n{}", idx + 1, backtrace));
+
+        Emit::Defer(format!(
+            "backtrace with {} frames ({})",
+            backtrace.frames().len(),
+            idx + 1
+        ))
+    }
+
+    #[cfg(feature = "spantrace")]
+    fn span_trace(spantrace: &SpanTrace, ctx: &mut HookContext<SpanTrace>) -> Emit {
+        let idx = ctx.increment();
+
+        let mut span = 0;
+        spantrace.with_spans(|_, _| {
+            span += 1;
+            true
+        });
+
+        ctx.attach_snippet(format!("Span Trace No. {}\n{}", idx + 1, spantrace));
+
+        Emit::Defer(format!("spantrace with {span} frames ({})", idx + 1))
+    }
+
+    /// Builtin hooks
+    ///
+    /// This provides defaults for common attachments that are automatically created
+    /// by `error_stack`, like [`Backtrace`] and [`SpanTrace`]
+    ///
+    /// [`Backtrace`]: std::backtrace::Backtrace
+    /// [`SpanTrace`]: tracing_error::SpanTrace
+    // Frame can be unused, if neither backtrace or spantrace are enabled
+    #[allow(unused_variables)]
+    // false positive
+    #[allow(clippy::missing_const_for_fn)]
+    pub fn builtin<'a>(frame: &Frame, ctx: &mut HookContext<'a, Frame>) -> Option<Emit> {
```

Comment:
As it's called `error_stack::fmt::builtin` this could be misleading (especially when adding other hooks later on.

```suggestion
    /// Fallback for common attachments that are automatically created
    /// by `error_stack`, like [`Backtrace`] and [`SpanTrace`]
    ///
    /// [`Backtrace`]: std::backtrace::Backtrace
    /// [`SpanTrace`]: tracing_error::SpanTrace
    // Frame can be unused, if neither backtrace or spantrace are enabled
    #[allow(unused_variables)]
    pub fn builtin_debug_hook_fallback<'a>(frame: &Frame, ctx: &mut HookContext<'a, Frame>) -> Option<Emit> {
```

I think the name describes it better. Even though it's a pretty long name I don't expect it to be used much, and if it's used, it's only used once per application (probably).